packages/api/src/controllers/experiment/attestation.ts

```diff
@@ -50,15 +58,64 @@ app.post("/", validatePost("attestation"), async (req, res) => {
   return res.status(201).json(attestationMetadata);
 });
 
-function verifySigner(message, signature): boolean {
-  const verifiedSigner = ethers.verifyTypedData(
-    DOMAIN,
-    TYPES,
-    message,
-    signature
-  );
+async function verifySignature(
+  message: any,
+  signature: string,
+  signatureType?: string
+): Promise<string> {
+  const verifiedEIP712 =
+    (signatureType === "eip712" || !signatureType) &&
+    verifyEIP712Signature(message, signature);
+  const verifiedFlow =
+    (signatureType === "flow" || !signatureType) &&
+    (await verifyFlowSignature(message, signature));
+
+  return verifiedEIP712 ? "eip712" : verifiedFlow ? "flow" : null;
+}
+
+function verifyEIP712Signature(message: any, signature: string): boolean {
+  try {
+    const verifiedSigner = ethers.verifyTypedData(
+      DOMAIN,
+      TYPES,
+      message,
+      signature
+    );
+
+    return verifiedSigner === message.signer;
+  } catch (e) {
+    return false;
+  }
+}
 
-  return verifiedSigner === message.signer;
+async function verifyFlowSignature(
+  message: any,
+  signature: string
+): Promise<boolean> {
+  try {
+    // These parameters are required to use the fcl library, even though we don't use on-chain verification
+    await fcl.config({
+      "flow.network": "testnet",
+      "accessNode.api": "https://access-testnet.onflow.org",
+    });
+
+    const compSig = [
+      {
+        f_type: "CompositeSignature",
+        f_vsn: "1.0.0",
+        addr: message.signer,
+        keyId: 0,
+        signature: signature,
+      },
+    ];
+
+    return await fcl.AppUtils.verifyUserSignatures(
```

Comment:
Curious do you know why their verification logic is async?

Reply:
I think they (by default) check the signatures on-chain. That is why all this logic is async. However, we (currently) don't check anything on-chain. Just their library is done that way.